Add worker maintenance mode functionality to Edge3 provider UI

dheerajturaga opened the pull request:
- Add POST /edge_worker/ui/worker/{worker_name}/maintenance endpoint to request maintenance mode
- Add DELETE /edge_worker/ui/worker/{worker_name}/maintenance endpoint to exit maintenance mode
- Add maintenance request form with required comment field (max 1024 chars)
- Add maintenance state handling for all maintenance-related worker states
- Add proper error handling and user feedback for maintenance operations
- Add worker existence validation and database session management
- Remove authentication dependencies for UI maintenance endpoints
- Update frontend to show maintenance controls based on worker state
- Format maintenance comments with timestamps for audit trail

<img width="1918" height="677" alt="image" src="https://github.com/user-attachments/assets/35034a9e-c788-48d9-ba89-98ee9b6c3019" />

<img width="1912" height="661" alt="image" src="https://github.com/user-attachments/assets/51178403-986c-43ee-b933-dbe3c7978dc4" />

<img width="1917" height="697" alt="image" src="https://github.com/user-attachments/assets/3b249fb2-c43c-4488-b1bf-a51c09254f2e" />

<img width="1202" height="532" alt="image" src="https://github.com/user-attachments/assets/77b33c37-ecb8-48ea-812b-d33a45935142" />

Expect more beautification in additional PRs, this is bringing in base functionality

dheerajturaga committed:
Add worker maintenance mode functionality to Edge3 provider UI

  - Add POST /edge_worker/ui/worker/{worker_name}/maintenance endpoint to request maintenance mode
  - Add DELETE /edge_worker/ui/worker/{worker_name}/maintenance endpoint to exit maintenance mode
  - Add maintenance request form with required comment field (max 1024 chars)
  - Add maintenance state handling for all maintenance-related worker states
  - Add proper error handling and user feedback for maintenance operations
  - Add worker existence validation and database session management
  - Remove authentication dependencies for UI maintenance endpoints
  - Update frontend to show maintenance controls based on worker state
  - Format maintenance comments with timestamps for audit trail

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+    """Request body for maintenance operations."""
+
+    maintenance_comment: Annotated[str, Field(description="Comment describing the maintenance reason.")]
+
```
For consistency, can you move this class to providers/edge3/src/airflow/providers/edge3/worker_api/datamodels_ui.py please?

dheerajturaga replied:
Done

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+
+@ui_router.post(
+    "/worker/{worker_name}/maintenance",
+)
```
To ensure this API is not "public/open" but authenticated, please add the access check
```suggestion
    dependencies=[
        Depends(requires_access_view(access_view=AccessView.JOBS)),
    ],
)
```

dheerajturaga replied:
@jscheffl , I already tried this and I was not able to get the post and delete to work. I keep getting 401 not authorized and I was not able to figure out how to provide proper authorization.

jscheffl replied:
Good. But still needed. Probably because you implemented the manual authentication in the service accessor.
Can you attempt it via the axios generated code? If then not working, please push and I can help fixing.

jscheffl replied:
Just added this in commit https://github.com/apache/airflow/pull/55348/commits/ec77bd65b96fd27eb601aeda51d4fb92354bc301 and it worked for me.

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+
+@ui_router.delete(
+    "/worker/{worker_name}/maintenance",
+)
```
Same as above
```suggestion
    dependencies=[
        Depends(requires_access_view(access_view=AccessView.JOBS)),
    ],
)
```

jscheffl replied:
Same here: Adding in commit https://github.com/apache/airflow/pull/55348/commits/ec77bd65b96fd27eb601aeda51d4fb92354bc301 and worked fine.

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+
+    try:
+        request_maintenance(worker_name, formatted_comment, session=session)
+        session.commit()  # Explicitly commit the transaction
```
commit is not needed, if no exception raised the ORM will commit() after return
```suggestion
```

dheerajturaga replied:
Done!

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+
+    try:
+        exit_maintenance(worker_name, session=session)
+        session.commit()  # Explicitly commit the transaction
```
as above
```suggestion
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+        request_maintenance(worker_name, formatted_comment, session=session)
+        session.commit()  # Explicitly commit the transaction
+    except Exception as e:
+        session.rollback()  # Rollback on error
```
As far as I see it implemented in other areas no rollback is needed if an exception is raised. ORM will roll back automatically.
```suggestion
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+        exit_maintenance(worker_name, session=session)
+        session.commit()  # Explicitly commit the transaction
+    except Exception as e:
+        session.rollback()  # Rollback on error
```
As above
```suggestion
```

dheerajturaga replied:
Done

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+    session: SessionDep,
+) -> None:
+    """Put a worker into maintenance mode."""
+    # Check if worker exists first
+    worker_query = select(EdgeWorkerModel).where(EdgeWorkerModel.worker_name == worker_name)
+    worker = session.scalar(worker_query)
+    if not worker:
+        raise HTTPException(status_code=404, detail=f"Worker {worker_name} not found")
```
Probably as follow-up... (no urgency to have it in this PR) the logic should be pushed into request_maintenance() method as it is redundant in CLI and other logic as well.

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/pages/WorkerPage.tsx:
```diff
@@ -16,22 +16,178 @@
+      console.error("Error exiting maintenance:", error);
+      alert(`Error exiting maintenance: ${error}`);
+    }
+  };
```
Actually I would assume you can use and leverage the generated axios wrappers and there is no need to implement the request and authentication manually.
See generated code in `services.gen.ts` and `queries.ts` -->`useUiServiceRequestWorkerMaintenance`

dheerajturaga replied:
Done!

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/pages/WorkerPage.tsx:
```diff
@@ -16,22 +16,178 @@
+    }
+
+    return null;
+  };
```
This code part looks like it would be better extracted into a React component so that it can be directly plugged into the table cell like `<OperationsCell worker={ worker } />`

dheerajturaga replied:
Done!

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +105,57 @@ def jobs(
+        raise HTTPException(status_code=404, detail=f"Worker {worker_name} not found")
+
+    # Format the comment with timestamp and username (username will be added by plugin layer)
+    formatted_comment = f"[{datetime.now().strftime('%Y-%m-%d %H:%M')}] - UI user put node into maintenance mode\nComment: {maintenance_request.maintenance_comment}"
```
Can you fetch the user name from the `depends` (or add a user depends as FastAPI decorator) and make this into the formatted string such that it is directly visible who made it to maintenance?

jscheffl replied:
Replication of commit https://github.com/apache/airflow/commit/ec77bd65b96fd27eb601aeda51d4fb92354bc301 (4)
```suggestion
    formatted_comment = f"[{datetime.now().strftime('%Y-%m-%d %H:%M')}] - {user.get_name()} put node into maintenance mode\nComment: {maintenance_request.maintenance_comment}"
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/pages/WorkerPage.tsx:
```diff
@@ -16,22 +16,178 @@
+      }
+      return (
+        <Button size="sm" colorScheme="blue" onClick={() => setActiveMaintenanceForm(workerName)}>
+          Enter Maintenance
```
Instead of (like in the 2.x legacy UI) rendering buttons with text - have you considered we use icons like in other tables/lists in the core UI?
Like here, icons on the right side:
<img width="1036" height="326" alt="image" src="https://github.com/user-attachments/assets/3c9226b9-2c96-407a-96a0-bf05b52474ef" />


For maintenance the icon https://react-icons.github.io/react-icons/search/#q=HiOutlineWrenchScrewdriver might be suitable (using this already for the state display)

jscheffl replied:
To exit maintenance e.g. the icon HiOutlineArrowRightStartOnRectangle or vsc/VscDebugRestart could be suitable

dheerajturaga replied:
I've updated to use the icons but I was not able to get tooltips working. It seems causing a Minified React Error 130

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/pages/WorkerPage.tsx:
```diff
@@ -16,22 +16,178 @@
+  };
+
+  return (
+    <VStack gap={2} align="stretch">
```
As with react we have much more cool widgets and options compared to legacy 2.x UI, how about making this a modal dialog?
Can also be extracted as a component into a separate tsx file to be used as `<MaintenanceForm worker={ worker } />`

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/pages/WorkerPage.tsx:
```diff
@@ -16,22 +16,178 @@
+          <Box fontSize="sm" whiteSpace="pre-wrap">
+            {worker.maintenance_comments || "No comment"}
+          </Box>
+          <Button size="sm" colorScheme="blue" onClick={() => exitMaintenance(workerName)}>
```
If we make this UI form now "right" I assume we should add some kind of confirmation dialog/modal such that the user needs to confirm via a "are you sure you want to exit maintenance for worker XYZ? (yes/no)"

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/pages/WorkerPage.tsx:
```diff
@@ -16,22 +16,178 @@
+      state === "maintenance pending" ||
+      state === "maintenance mode" ||
+      state === "maintenance request" ||
+      state === "maintenance exit" ||
```
I think when maintenance exit is already there as status we do not need to present the "Exit" button again, it is just pending that the worker fetches this.
```suggestion
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/pages/WorkerPage.tsx:
```diff
@@ -16,22 +16,178 @@
+
+  return (
+    <VStack gap={2} align="stretch">
+      <Textarea
```
Would be a cool improvement over 2.x UI to be able to use Markdown rendering here? WDYT? We might need to clone the component from core-ui. Might be also a beautification as a follow-up PR.

jscheffl reviewed (commented):
Very cool! I also wanted to get started with the maintenance but you opened the PR faster. Cool! 
Functionally the skeleton is working.

Sorry about the many comments, don't be afraid. If too much then let me know and I can also contribute to this as this is an important feature urgently needed in UI for Airflow 3.1 release.

dheerajturaga committed:
Move MaintenanceRequest to datamodels_ui.py for better code organization
Remove explicit session.commit() calls - let SQLAlchemy handle transactions

dheerajturaga committed:
refactor: extract OperationsCell into separate component

  - Move MaintenanceRequest class to datamodels_ui.py for better organization
  - Remove explicit session.commit() calls, let SQLAlchemy handle transactions
  - Extract OperationsCell and MaintenanceForm into separate component file
  - Clean up unused imports in WorkerPage.tsx
  - Improve code modularity and maintainability

dheerajturaga committed:
refactor: replace manual fetch with generated axios wrappers for maintenance operations

  - Use useUiServiceRequestWorkerMaintenance and useUiServiceExitWorkerMaintenance hooks
  - Remove manual fetch implementation with CSRF token handling
  - Simplify error handling using React Query's onSuccess/onError callbacks
  - Improve type safety with generated API client
  - Reduce bundle size and code complexity
  - Maintain same functionality with cleaner, more maintainable code

dheerajturaga committed:
Use react icons instead of old school buttons

jscheffl committed:
Add tooltips to buttons

jscheffl committed:
Have exit button not if it is in maintenance exit already

jscheffl committed:
Add authentication to (new) API endpoints

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/components/OperationsCell.tsx:
```diff
@@ -0,0 +1,138 @@
+        >
+          <FcCheckmark />
+        </IconButton>
+        <IconButton size="sm" colorScheme="red" variant="outline" onClick={onCancel} aria-label="Cancel">
```
Replication of commit https://github.com/apache/airflow/pull/55348/commits/fdd0c2b0f7eeeda79a8e00cec50a7417ef0f9406 (1)
```suggestion
        <IconButton
          size="sm"
          colorScheme="red"
          variant="outline"
          onClick={onCancel}
          aria-label="Cancel"
          title="Cancel"
        >
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/components/OperationsCell.tsx:
```diff
@@ -0,0 +1,138 @@
+            size="sm"
+            variant="ghost"
+            onClick={() => onSetActiveMaintenanceForm(workerName)}
+            aria-label="Enter Maintenance"
```
Replication of commit https://github.com/apache/airflow/commit/fdd0c2b0f7eeeda79a8e00cec50a7417ef0f9406 (2)
```suggestion
            aria-label="Enter Maintenance"
            title="Enter Maintenance"
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/components/OperationsCell.tsx:
```diff
@@ -0,0 +1,138 @@
+            size="sm"
+            variant="ghost"
+            onClick={() => onExitMaintenance(workerName)}
+            aria-label="Exit Maintenance"
```
Replication of commit https://github.com/apache/airflow/commit/fdd0c2b0f7eeeda79a8e00cec50a7417ef0f9406 (3)
```suggestion
            aria-label="Exit Maintenance"
            title="Exit Maintenance"
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/plugins/www/src/components/OperationsCell.tsx:
```diff
@@ -0,0 +1,138 @@
+    state === "maintenance pending" ||
+    state === "maintenance mode" ||
+    state === "maintenance request" ||
+    state === "maintenance exit" ||
```
Replication of commit https://github.com/apache/airflow/pull/55348/commits/119ad58cecb4155a23a88783bfb6e96ea26ae37c
```suggestion
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -17,18 +17,21 @@
 from airflow.api_fastapi.auth.managers.models.resource_details import AccessView
 from airflow.api_fastapi.common.db.common import SessionDep  # noqa: TC001
 from airflow.api_fastapi.common.router import AirflowRouter
 from airflow.api_fastapi.core_api.security import requires_access_view
```
Replication of commit https://github.com/apache/airflow/pull/55348/commits/ec77bd65b96fd27eb601aeda51d4fb92354bc301 (1)
```suggestion
from airflow.api_fastapi.core_api.security import GetUserDep, requires_access_view
```

jscheffl commented on providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py:
```diff
@@ -100,3 +103,47 @@ def jobs(
+def request_worker_maintenance(
+    worker_name: str,
+    maintenance_request: MaintenanceRequest,
+    session: SessionDep,
```
Replication of commit https://github.com/apache/airflow/pull/55348/commits/ec77bd65b96fd27eb601aeda51d4fb92354bc301 (3)
```suggestion
    session: SessionDep,
    user: GetUserDep,
```

jscheffl reviewed (commented):
I was not able to commit to your repo, Github gave me the following error :-(
```
jscheffl@hp860g9:~/Workspace/airflow_review$ git push
error: Authentication error: Authentication required: You must have push access to verify locks
error: failed to push some refs to 'github.com:dheerajturaga/airflow.git'
```

Therefore (1) added commits 1-3 from my additions as recommondations/comments. Did not pply them because generated and compiled code was not possible to be added by this ... you need to apply and compile UI.

Alternatively opened a DRAFT PR, you can also easily `git cherry-pick` from there:
https://github.com/apache/airflow/pull/55348:
1. [Add tooltips to buttons](https://github.com/apache/airflow/pull/55348/commits/fdd0c2b0f7eeeda79a8e00cec50a7417ef0f9406)
2. [Have exit button not if it is in maintenance exit already](https://github.com/apache/airflow/pull/55348/commits/119ad58cecb4155a23a88783bfb6e96ea26ae37c)
3. [Add authentication to (new) API endpoints](https://github.com/apache/airflow/pull/55348/commits/ec77bd65b96fd27eb601aeda51d4fb92354bc301)

My change attempt to make a proper dialog in commit 4 / [Rework maintenance dialog to be a dialog](https://github.com/apache/airflow/pull/55348/commits/7e6d13cf316711f286e3262fa8552b2d62ede965) as not successful, asked in Slack if I am tooo stupid... I assume to need help on this.
So to un-block the efforts, maybe if commits 1-3 are cherry picked it is OK to merge and then applying beautification later?

jscheffl reviewed (approved):
Looks good!
